main/lsp/LSPTypechecker.cc

```diff
@@ -0,0 +1,537 @@
+}
+
+void LSPTypechecker::initialize(LSPFileUpdates updates, LSPOutput &output) {
+    runOnTypecheckerThread([&]() -> void {
+        globalStateHashes = move(updates.updatedFileHashes);
+        indexed = move(updates.updatedFileIndexes);
+        // Initialization typecheck is not cancelable.
+        auto run = runSlowPath(move(updates), /* cancelable */ false, output);
+        ENFORCE(!run.canceled);
+        ENFORCE(run.newGS.has_value());
+        gs = move(run.newGS.value());
+        pushDiagnostics(move(run), output);
+        config.initialized = true;
+    });
+}
+
+bool LSPTypechecker::typecheck(LSPFileUpdates updates, LSPOutput &output) {
+    bool finished = true;
+    // TODO: Make this asynchronous.
+    runOnTypecheckerThread([&]() -> void {
+        auto run = runTypechecking(move(updates), output);
+        finished = !run.canceled;
+        commitTypecheckRun(move(run), output);
+    });
+    return finished;
+}
+
+void LSPTypechecker::enterCriticalSection(std::function<void(const LSPTypecheckerOps &)> lambda) {
+    runOnTypecheckerThread([&]() -> void { lambda(LSPTypecheckerOps(*this, *gs)); });
+}
+
+void LSPTypechecker::shutdown() {
+    absl::MutexLock lock(&mtx);
+    wasShutdown = true;
+}
+
+unique_ptr<Joinable> LSPTypechecker::runTypechecker() {
+    ENFORCE(this_thread::get_id() == typecheckerThreadId);
+    return runInAThread("lspPreprocess", [&]() -> void {
+        typecheckerThreadId = this_thread::get_id();
+        optional<function<void()>> toRun = nullopt;
+        {
+            absl::MutexLock lock(&mtx);
+            mtx.Await(absl::Condition(
+                +[](LSPTypechecker *typechecker) -> bool {
+                    typechecker->mtx.AssertHeld(); // Satisfy thread annotations.
+                    return typechecker->wasShutdown || typechecker->lambda.has_value();
+                },
+                this));
+            if (wasShutdown) {
+                lambda = nullopt;
+                return;
+            }
+            toRun = move(lambda);
+            lambda = nullopt;
+        }
+        ENFORCE(toRun.has_value());
+        toRun.value()();
+        toRun = nullopt;
+    });
+}
+
+void LSPTypechecker::runOnTypecheckerThread(function<void()> lambda) {
+    if (this_thread::get_id() == typecheckerThreadId) {
+        return lambda();
+    } else {
+        // TODO: Support async tasks. For now, everything blocks.
+        absl::Notification lambdaComplete;
+        {
+            absl::MutexLock lock(&mtx);
+            ENFORCE(!this->lambda.has_value());
+            this->lambda = [&lambdaComplete, &lambda]() -> void {
+                lambda();
+                lambdaComplete.Notify();
+            };
+        }
+        // Block until lambda runs.
+        lambdaComplete.WaitForNotification();
+    }
+}
+
+TypecheckRun LSPTypechecker::runTypechecking(LSPFileUpdates updates, LSPOutput &output) const {
```

Comment:
Can we separate the methods related to coordination&mutithreading from ones related to typechecking into different classes and use delegation?

Reply:
Good idea; that will be much cleaner.